src/main/scala/com/github/codelionx/dodo/actors/ODMaster.scala

```diff
@@ -296,116 +293,189 @@ class ODMaster(inputFile: Option[File])
+              None
+            }
         }
+        context.become(workStealing(table, newPendingResponses.toSet))
       } else if (candidateQueue.hasNoPendingWork) {
-        // we have no work left and got no work from our work stealing attempt, finished?
-        // TODO: make sure others don't have anything pending work anymore as well
-        log.warning(
-          "This node has no more pending candidates and the queue of other nodes is empty as well. Shutting down node."
-        )
-        shutdown()
+        // we have no work left and got no work from our work stealing attempt
+        if (pendingSum > 0) {
+          // others will soon have work for us to steal again
+          log.info("No work received, will ask again in three seconds")
+          context.system.scheduler.scheduleOnce(3 second, self, startWorkStealing(table))
+        }
+        // check if ALL nodes are completely out of work
+        startDowningProtocol(table)
       } else {
         // nothing to steal, but we still wait for worker results, wait for them
         log.info("Nothing to steal, but our workers are still busy.")
         context.become(findingODs(table))
       }
 
+    case Terminated(otherMaster) =>
+      val newPendingResponses = pendingResponses - otherMaster
+      context.become(workStealing(table, newPendingResponses))
+      if (newPendingResponses.isEmpty) {
+        if (candidateQueue.queueSize == 0 && candidateQueue.pendingSize == 0) {
+          log.info("Work stealing was unsuccessful")
+          startDowningProtocol(table)
+        } else {
+          context.become(findingODs(table))
+        }
+      }
+
     case StolenWork(stolenQueue) =>
       log.info("Received work from {}", sender)
       candidateQueue.enqueue(stolenQueue)
       sender ! AckWorkReceived
       sendWorkToIdleWorkers()
-      context.become(findingODs(table))
+      val newPendingResponses = pendingResponses - sender
+      context.unwatch(sender)
+      if (newPendingResponses.isEmpty) {
+        if (candidateQueue.queueSize == 0 && candidateQueue.pendingSize == 0) {
+          log.info("Work stealing was unsuccessful")
+          startDowningProtocol(table)
+        } else {
+          context.become(findingODs(table))
+        }
+      } else {
+        context.become(workStealing(table, newPendingResponses))
+      }
 
     case ODsToCheck(newODs) =>
       candidateQueue.enqueueNewAndAck(newODs, sender)
       sendWorkToIdleWorkers()
 
-    case GetWorkLoad if sender == self => // ignore
-    case ReportReducedColumnStatus => // ignore
+    case WorkToSend(amount: Int) =>
+      log.info("No work to send")
+      sender ! StolenWork(Queue.empty)
 
+    case ReportReducedColumnStatus => // ignore
     case m => log.debug("Unknown message received in `workStealing`: {}", m)
   }
 
+  def downing(table: Array[TypedColumn[Any]], pendingMasters: Set[Address]): Receive = withGetWorkLoadHandling {
+    case CurrentClusterState(members) =>
```

Comment:
Can you destructure the `members` already in the case matching, so we know what `_3` means?

Would look similar to this:
```scala
case CurrentClusterState((_, _, nameFor_3, ...)) =>
```